### PR TITLE
Build & publish images through GHA

### PR DESCRIPTION
**Pull request by charlesbluca**
Working on refactoring the existing infra in this repo to build and publish through a GHA run, which theoretically should give us a lot more flexibility in automating / publishing status checks of builds.

### CODE REVIEW

File: .github/workflows/build.yml
Comment (charlesbluca) [inline .github/workflows/build.yml:10]: Not sure if there's a preference on how something like this would get triggered on a schedule? Ideally would like to have the status checks broadcast on Slack to improve visibility, so happy to move in whatever direction makes that easiest with our current devops stack

File: .github/workflows/build.yml
Comment (charlesbluca) [inline .github/workflows/build.yml:47]: These are just filler for now, assuming these don't exist - @ajschmidt8 are there org/repo secrets we could use here to authenticate with the `gpuci` dockerhub?

https://hub.docker.com/u/gpuci

File: dask/Dockerfile
Comment (charlesbluca) [inline dask/Dockerfile:16]: Is this an overly aggressive merging of these layers? The initial problem this aimed to solve was storage pressure around making the conda install and `chmod`ing of the conda files separate layers

**Review by charlesbluca — COMMENTED** [review]
One question that isn't really contained in these code changes - the current docker build commands we're using to make the images use `--squash`, is there any way to mimic that through the `docker/build-push-action` step?

File: dask/Dockerfile
Comment (AyodeAwe) [reply dask/Dockerfile:16]: Might want to consider using Dockerfile `heredocs`?

- https://www.docker.com/blog/introduction-to-heredocs-in-dockerfiles/
- https://docs.docker.com/engine/reference/builder/#example-running-a-multi-line-script

File: .github/workflows/build.yml
Comment (AyodeAwe) [reply .github/workflows/build.yml:10]: This looks ok to me. cc @raydouglass for additional review

FWIW this post might be relevant for your slack notification needs: https://github.blog/changelog/2022-12-06-github-actions-workflow-notifications-in-slack-and-microsoft-teams/

File: .github/workflows/build.yml
Comment (ajschmidt8) [reply .github/workflows/build.yml:47]: I would check with @raydouglass. I think that DockerHub organization is largely defunct except for the Dask/Distributed images.

Ray will have guidance on whether you should keep using it. And if so, he can also provide credentials.

File: dask/Dockerfile
Comment (charlesbluca) [reply dask/Dockerfile]: Thanks for the suggestion! Done with https://github.com/rapidsai/dask-build-environment/pull/105/commits/d1141c1d6ab4e4d1cd35a898ca1ba969ec47abc1 - https://github.com/rapidsai/dask-build-environment/pull/105/commits/4a0f0f37d504897cea119338f985ab2f4f3618e2

File: .github/workflows/update-gpuci.yml
Comment (jameslamb) [inline .github/workflows/update-gpuci.yml:22]: This was the only use of `ci/axis.dask.yaml`.

```shell
git grep axis
```

That file should be deleted as part of this PR.

File: .github/workflows/update-gpuci.yml
Comment (jameslamb) [inline .github/workflows/update-gpuci.yml:27]: I'm uncomfortable with implicitly depending on the ordering of the `matrix.rapids` entry here. I could see someone changing `["24.10", "24.12"]` to `["25.02", "24.12"]` in a PR and not realizing that that was breaking some of the assumptions here.

I also think it's wasteful (and a source of unnecessary risk of failed builds due to network issues) to have all these calls to `https://version.gpuci.io/rapids`. The response from that endpoint should change only when the RAPIDS version changes. And if you're only tracking 2 RAPIDS versions, I think it'd be simpler to just hard-code the corresponding `ucx-py` version.

Would you consider something like the following over in `build.yml`? At the top level of the workflow:

```yaml
env:
  OLD_RAPIDS_VER: "24.10"
  OLD_UCX_PY_VER: "0.40"
  RAPIDS_VER: "24.12"
  UCX_PY_VER: "0.41"

jobs:
  docker:
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        cuda: ["11.8.0"]
        python: ["3.10", "3.11", "3.12"]
        linux: ["ubuntu20.04"]
        rapids: [${{ env.OLD_RAPIDS_VER }}, ${{ env.NEW_RAPIDS_VER }}]
```

And then passing those values as outputs through to whatever needs it in this `update-gpuci.yml`? That'd remove that source of network calls, and remove this implicit reliance on the ordering of the matrix entries for correctness.

Comment (jameslamb) [reply .github/workflows/update-gpuci.yml:27]: I think that'd also allow you to get rid of all this other machinery:

https://github.com/rapidsai/dask-build-environment/blob/89770692cee67f261d0a5cdd5f5b28ef82342450/.github/workflows/update-gpuci.yml#L38-L44

There's exactly 1 `ucx-py` version per RAPIDS release and these builds seem to assume they'll only run for exactly 2 RAPIDS versions... seems like a great opportunity to replace some complicated and custom machinery with a little bit of hard-coding that'll be updated infrequently.

File: .github/workflows/update-gpuci.yml
Comment (jameslamb) [inline .github/workflows/update-gpuci.yml:57]: Why are these `ci/**` filters no longer necessary? I think *some* filters should be kept, assuming I understand how the templating is working in this repo. Otherwise, you risk modifying things like these `.github/workflows` files, which might lead to surprising results (I don't know, for example, how and when GitHub Actions parses these files).

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this! I left some comments for your consideration. I think there are some opportunities to significantly simplify this. Please let me know if you'd like to pair on that.

File: .github/workflows/build.yml
Comment (charlesbluca) [reply .github/workflows/build.yml:85]: Yeah this definitely makes more sense to include as part of the build matrix - will add this up there

File: .github/workflows/update-gpuci.yml
Comment (charlesbluca) [reply .github/workflows/update-gpuci.yml:22]: Probably should've provided more clarity around this in the PR description, but this file is actually getting pointed to by some of our Jenkins infra, which I'm assuming would start failing if we deleted this file:

https://github.com/rapidsai/gpuci-scripts/blob/aef3a3321f0393b2473214235b0c7b6e3950b1db/seed-job-dsl/jobs-other/dask/dask-build-environment/branch/dask_build_env_main.groovy#L1

My thinking was that we would:

- merge this in, ensure that the GHA image publishing is working as expected alongside the Jenkins job
- make the necessary internal changes to drop the scheduled Jenkins runs, which should allow us to delete the above linked file
- drop the contents of `ci/**` in this repo as a follow up once the Jenkins runs have been turned off

Does that seem like a reasonable approach? Happy to do things a different way, and/or add more writing here to clarify the other ops work that should be considered a follow up to this PR.

File: .github/workflows/update-gpuci.yml
Comment (charlesbluca) [reply .github/workflows/update-gpuci.yml:27]: Yeah I definitely think the ordering concern alone is sufficient to hardcode these values somewhere other than the build matrix itself, shouldn't be too difficult to refactor the builds to rely on an internal version mapping instead of `https://version.gpuci.io/rapids`

> I think that'd also allow you to get rid of all this other machinery

Were you thinking of replacing this step with a request to `https://version.gpuci.io/rapids` to fetch the new UCX-Py version instead? That could work, IIRC I opted to do it this way because cuDF and UCX-Py nightlies became available at different times in the RAPIDS release process, and so this ensured that both were available before opening a PR that would inevitably fail at the conda solve step (though this may be less of a problem now with PR checks running to make PRs like that visibly "blocked" from merging)

Comment (charlesbluca) [reply .github/workflows/update-gpuci.yml]: Another minor annoyance here I didn't foresee - it looks like the `env` context is not available within the `matrix` block, so the example you shared above triggers a syntax error:

https://github.com/rapidsai/dask-build-environment/actions/runs/11114625627/workflow

File: .github/workflows/update-gpuci.yml
Comment (charlesbluca) [reply .github/workflows/update-gpuci.yml:57]: Done in https://github.com/rapidsai/dask-build-environment/pull/105/commits/fec0a76360a86e3e61d53cca0a0d1da80741eb94

File: .github/workflows/update-gpuci.yml
Comment (charlesbluca) [reply .github/workflows/update-gpuci.yml]: Was able to get around this by using some hardcoded representative values in the build matrix, that we then later manually map to the hardcoded RAPIDS/ucx-py versions in https://github.com/rapidsai/dask-build-environment/pull/105/commits/0af7f24e726b2170de32201e1518fc3c5d26449a

File: .github/workflows/update-gpuci.yml
Comment (jameslamb) [reply .github/workflows/update-gpuci.yml:22]: ooooo thank you for that explanation! Yes I totally support that plan, I'd forgotten about the Jenkins half of this.

File: .github/workflows/update-gpuci.yml
Comment (jameslamb) [reply .github/workflows/update-gpuci.yml]: > *were you thinking of replacing this step*

I looked back through this workflow more closely to be sure I understood what it was doing. I didn't understand that it's job was to trigger new PRs here, so some of my earlier comments about it might not make sense.

Am I right that this `update-gpuci` workflow's role is "open a PR here in `dask-build-environment` whenever a new RAPIDS release or `ucx-py` release is detected?".

https://github.com/rapidsai/dask-build-environment/blob/89770692cee67f261d0a5cdd5f5b28ef82342450/.github/workflows/update-gpuci.yml#L86-L98

Like #104?

If so... I'd just remove this entire `update-gpuci.yml` workflow. Maybe things were different back when it was introduced in #10, but today the amount of manual work is so small here and happens so infrequently, I just don't think all this complexity in GitHub Actions is worth it.

There's a new RAPIDS release once every 2 months, and every RAPIDS release corresponds to exactly one `ucx-py` release.

Specifically, I'm proposing:

* remove `update-gpuci.yml` workflow completely
* ask RAPIDS ops if updating this repo can be a part of the RAPIDS release checklist
   - *(optional): automate those updates via a `ci/release/update-version.sh`, like most other RAPIDS repos do, e.g. https://github.com/rapidsai/docker/blob/branch-24.10/ci/release/update-version.sh*

If you want to keep this automated, then ignore my other comments about `update-gpuci` and leave the network calls to get those versions as-is.

Comment (jameslamb) [reply .github/workflows/update-gpuci.yml]: > *looks like the env context is not available within the matrix block*

blegh ok, sorry for misleading you there. I like the solution you came up with!

**Review by jameslamb — APPROVED** [review]
Now that I understand it more deeply, I really think that the `update-gpuci.yml` workflow should just be removed completely, in favor of manual action once every 2 months (each time there's a new RAPIDS release).

But that conversation doesn't need to block this PR. Changes look good to me, thanks for working through some of my suggestions.

**Comment by charlesbluca** [comment]
Would've thought the approval from @jameslamb would be sufficient to unblock merging, but it looks he isn't a member of the `rapidsai/dask-build-environment-write` team - bumped https://github.com/rapidsai/ops/issues/3481 to see if we want to add more people to this list

File: .github/workflows/update-gpuci.yml
Comment (charlesbluca) [reply .github/workflows/update-gpuci.yml]: > Am I right that this update-gpuci workflow's role is "open a PR here in dask-build-environment whenever a new RAPIDS release or ucx-py release is detected?".

Pretty much! Only thing I'd add is that it's specifically when a new _cuDF_ release _and_ ucx-py release are detected, not if it's just or the other.

> ask RAPIDS ops if updating this repo can be a part of the RAPIDS release checklist

Definitely a strong +1 on adding something like this to a checklist to increase the visibility of the ops work around Dask GPU CI, worth noting that there are similar workflows in the respective Dask repos to automate this version bump too:

- https://github.com/dask/dask/blob/main/.github/workflows/update-gpuci.yml
- https://github.com/dask/distributed/blob/main/.github/workflows/update-gpuci.yaml

Which reflect other steps that could probably be done manually as part of a checklist - maybe we should file an issue here to track this switch?

**Review by rjzamora — APPROVED** [review]
Approving the general direction of this PR. Deferring to @jameslamb 's technical approval.

File: .github/workflows/update-gpuci.yml
Comment (jameslamb) [reply .github/workflows/update-gpuci.yml]: Sure! Up to you and the other Dask folks how you want to handle it. Also fine to tell me "nah we'll keep this as-is", I don't feel that strongly about it. Just right now, in this moment, this feels like a lot of complexity to automate something that changes very infrequently.

**Pull request merged**
